smarts/core/utils/traffic_history_serivce.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+from dataclasses import dataclass
+from multiprocessing import Pipe, Process, Queue
+
+import ijson
+
+import smarts.core.scenario as scenario
+
+
+@dataclass
+class RequestHistoryRange:
+    start_index: int
+    batch_count: int
+
+
+class Traffic_history_service:
+    """responsible for dynamically fetching traffic history json to reduce
+    memory use of traffic history data
+    """
+
+    def __init__(self, history_file_path):
+        self._history_file_path = history_file_path
+        self._all_timesteps = set()
+        self._current_traffic_history = {}
+        self._prev_batch_history = {}
+        # return if traffic history is not used
+        if history_file_path is None:
+            return
+
+        send_data_conn, receive_data_conn = Pipe()
+        self._receive_data_conn = receive_data_conn
+        self._request_queue = Queue()
+        self._fetch_history_proc = Process(
+            target=self._fetch_history,
+            args=(
+                send_data_conn,
+                self._request_queue,
+                self._history_file_path,
+            ),
+        )
+        self._fetch_history_proc.daemon = True
+        self._fetch_history_proc.start()
+
+        self._range_start = 0
+        self._batch_size = 300
+        # initialize
+        with open(self._history_file_path, "rb") as f:
+            for index, (t, vehicles_state) in enumerate(
+                ijson.kvitems(f, "", use_float=True)
+            ):
+                self._all_timesteps.add(t)
+                if (
+                    self._range_start <= index
+                    and index < self._range_start + self._batch_size
+                ):
+                    self._current_traffic_history[t] = vehicles_state
```

Comment:
Are the iterations of the `for` loop independent, or can be made independent, such that they can be replaced with `map` or other constructs to improve performance and speed?

Reply:
for a long block of code like this, is it suitable to use `map`? Since lambda functions can only be 1 line, I would have to define a new function for the following code:
```
self._all_timesteps.add(t)
if (
    self._range_start <= index
    and index < self._range_start + self._batch_size
):
    self._current_traffic_history[t] = vehicles_state
```

Reply:
To avoid repetitions, the following comment is applicable to all of my previous comments on trying to avoid `for` loops.
 
A good practise would be to avoid executing code serially in a `for` loop if the contents of the `for` loop is independent of the iterations. Unnecessary `for` loops could eventually build up to slow down the code. It would be difficult to refactor the code at a later time, more so by a different author. Hence, it is mostly best that the code is well parallelized the first time it is written. If needed, additional functions may be written for usage in `map` or other constructs.

However, usage of `for` loops may be acceptable in cases where the `for` loop only minimally increases the run time or if the alternative parallel execution consumes more run time due to increase setup time.

You would be in the best position to determine whether the code can be refactored to avoid `for` loops to gain speed, if it is possible. If it is not possible to gain any further speed ups, than I am happy to continue merging this pull request and I have approved it.